token/js/src/extensions/transferHook/state.ts

```diff
@@ -49,3 +52,64 @@ export function getTransferHookAccount(account: Account): TransferHookAccount |
         return null;
     }
 }
+
+/** ExtraAccountsMeta as stored by the transfer hook program */
+export interface ExtraAccountsMeta {
+    discriminator: number;
+    addressConfig: Uint8Array;
+    isSigner: boolean;
+    isWritable: boolean;
+}
+
+/** Buffer layout for de/serializing an ExtraAccountsMeta */
+export const ExtraAccountsMetaLayout = struct<ExtraAccountsMeta>([
+    u8('discriminator'),
+    blob(32, 'addressConfig'),
+    bool('isSigner'),
+    bool('isWritable'),
+]);
+
+export interface ExtraAccountsMetaList {
+    count: number;
+    extraAccounts: ExtraAccountsMeta[];
+}
+
+/** Buffer layout for de/serializing a list of ExtraAccountsMeta prefixed by a u32 length */
+export const ExtraAccountsMetaListLayout = struct<ExtraAccountsMetaList>([
+    u32('count'),
+    seq<ExtraAccountsMeta>(ExtraAccountsMetaLayout, greedy(ExtraAccountsMetaLayout.span), 'extraAccounts'),
+]);
+
+/** Unpack an extra account metas account and parse the data into a list of ExtraAccountMetas */
+export function getExtraAccountMetas(account: AccountInfo<Buffer> | null): ExtraAccountsMeta[] | null {
+    if (account == null) {
+        throw null;
+    }
+    return ExtraAccountsMetaListLayout.decode(account.data).extraAccounts;
+}
+
+/** Take an ExtraAccountMeta and construct that into an acutal AccountMeta */
+export function resolveExtraAccountMeta(
+    extraMeta: ExtraAccountsMeta,
+    previousMetas: AccountMeta[],
+    instructionData: Buffer,
+    transferHookProgramId: PublicKey
+): AccountMeta {
+    if (extraMeta.discriminator === 0) {
+        return {
+            pubkey: new PublicKey(extraMeta.addressConfig),
+            isSigner: extraMeta.isSigner,
+            isWritable: extraMeta.isWritable,
+        };
+    }
+
+    const accountIndex = extraMeta.discriminator - (1 << 7);
+    if (previousMetas.length <= accountIndex) {
+        throw new TokenTransferHookInvalidDiscriminator();
+    }
+    const programId = extraMeta.discriminator === 1 ? transferHookProgramId : previousMetas[accountIndex].pubkey;
```

Comment:
What do you think of not trying to evaluate the index for the account without first checking the discriminator?

```suggestion
    let programId;
    if (extraMeta.discriminator === 1) {
        programId = transferHookProgramId;
    } else {
        const accountIndex = extraMeta.discriminator - (1 << 7);
        if (previousMetas.length <= accountIndex) {
            throw new TokenTransferHookAccountNotFound(); // New error
        }
        programId = previousMetas[accountIndex].pubkey
    }
```

Note this uses a new error type - which matches the one in the Rust lib - for `AccountNotFound`

Reply:
Yep this makes more sense